Src/FluentAssertions/Common/Services.cs

```diff
@@ -33,14 +38,66 @@ public static Configuration Configuration
 
     public static IReflector Reflector { get; set; }
 
+    [PublicAPI]
     public static void ResetToDefaults()
     {
-        Reflector = new FullFrameworkReflector();
+        isInitialized = false;
+        EnsureInitialized();
+    }
+
+    internal static void EnsureInitialized()
+    {
+        if (isInitialized)
+        {
+            return;
+        }
+
+        lock (Lockable)
+        {
+            if (!isInitialized)
+            {
+                ExecuteCustomInitializers();
+
+                Reflector = new FullFrameworkReflector();
 #if NETFRAMEWORK || NET6_0_OR_GREATER
-        ConfigurationStore = new ConfigurationStoreExceptionInterceptor(new AppSettingsConfigurationStore());
+                ConfigurationStore = new ConfigurationStoreExceptionInterceptor(new AppSettingsConfigurationStore());
 #else
-        ConfigurationStore = new NullConfigurationStore();
+                    ConfigurationStore = new NullConfigurationStore();
 #endif
-        ThrowException = new TestFrameworkProvider(Configuration).Throw;
+                ThrowException = new TestFrameworkProvider(Configuration).Throw;
+
+                isInitialized = true;
+            }
+        }
+    }
+
+    private static void ExecuteCustomInitializers()
+    {
+        var currentAssembly = Assembly.GetExecutingAssembly();
+        var currentAssemblyName = currentAssembly.GetName();
+
+        try
+        {
+            var attributes = AppDomain.CurrentDomain
+                .GetAssemblies()
+                .Where(assembly => assembly != currentAssembly && !assembly.IsDynamic && !IsFramework(assembly))
+                .Where(a => a.GetReferencedAssemblies().Any(r => r.FullName == currentAssemblyName.FullName))
+                .SelectMany(a => a.GetCustomAttributes<AssertionEngineInitializerAttribute>());
```

Comment:
We might want to be more specific on the method signature we scan for - but this could also be improved in a later PR.

Even though we catch exceptions, it's annoying if a third party library throws internally and triggers a debugger to break execution.

This example throws `System.Reflection.TargetParameterCountException: 'Parameter count mismatch.'`
```cs
public static void InitializeBeforeFirstAssertion(object _)
{
}
```

This example throws `System.Reflection.AmbiguousMatchException: 'Ambiguous match found.'`
```cs
public static void InitializeBeforeFirstAssertion()
{
    Interlocked.Increment(ref shouldBeCalledOnlyOnce);
}

public static void InitializeBeforeFirstAssertion(object _)
{
}
```

Reply:
At this point, I'm not checking for any signatures. Am I missing something?

Reply:
Right, we don't filter on signatures here, i.e. at this point we don't try to validate if the method about the be called exists.
When invoking `attribute.Initialize()` and hence `AssertionEngineInitializerAttribute.Initialize` we assume that `type` has a single static parameterless method named `methodName`.

As this hopefully is an extreme edge case and nothing we can do at this point in the code the fix it, I'm fine with the added try/catch inside the loop to prevent bogus ones interfering with working ones.